tikv/raftstore/fsm_peer.go

```diff
@@ -636,23 +638,224 @@ func (d *peerFsmDelegate) findOverlapRegion(storeMeta *storeMeta, snapRegion *me
+		// so we send it independently here.
+		task := pd.Task{
+			Type: pd.TaskReportBatchSplit,
+			Data: &pd.ReportBatchSplit{Regions: regions},
+		}
+		d.ctx.pdScheduler <- task
+	}
+
+	lastRegion := regions[len(regions)-1]
+	if !meta.regionRanges.Delete(lastRegion.EndKey) {
+		panic(d.tag() + " original region should exist")
+	}
+	// It's not correct anymore, so set it to None to let split checker update it.
+	d.peer.ApproximateSize = nil
+	lastRegionID := lastRegion.Id
+	for _, newRegion := range regions {
+		newRegionID := newRegion.Id
+		value := make([]byte, 8)
+		binary.LittleEndian.PutUint64(value, newRegionID)
+		notExist := meta.regionRanges.Insert(newRegion.EndKey, value)
+		y.Assert(notExist)
+		if newRegionID == regionID {
+			continue
+		}
+
+		// Insert new regions and validation
+		log.Infof("[region %d] inserts new region %s", regionID, newRegion)
+		if r, ok := meta.regions[newRegionID]; ok {
+			// Suppose a new node is added by conf change and the snapshot comes slowly.
+			// Then, the region splits and the first vote message comes to the new node
+			// before the old snapshot, which will create an uninitialized peer on the
+			// store. After that, the old snapshot comes, followed with the last split
+			// proposal. After it's applied, the uninitialized peer will be met.
+			// We can remove this uninitialized peer directly.
+			if len(r.Peers) > 0 {
+				panic(fmt.Sprintf("[region %d] duplicated region %s for split region %s",
+					newRegionID, r, newRegion))
+			}
+			d.ctx.router.close(newRegionID)
+		}
+
+		sender, newPeer, err := createPeerFsm(d.ctx.store.Id, d.ctx.Cfg, d.ctx.regionScheduler, d.ctx.engine, newRegion)
+		if err != nil {
+			// peer information is already written into db, can't recover.
+			// there is probably a bug.
+			panic(fmt.Sprintf("create new split region %s error %v", newRegion, err))
+		}
+		metaPeer := ClonePeer(newPeer.peer.Peer)
+		for _, p := range newRegion.Peers {
+			// Add this peer to cache.
+			newPeer.peer.InsertPeerCache(ClonePeer(p))
+		}
+
+		// New peer derive write flow from parent region,
+		// this will be used by balance write flow.
+		newPeer.peer.PeerStat = d.peer.PeerStat
+		campaigned := newPeer.peer.MaybeCampaign(isLeader)
+		newPeer.hasReady = newPeer.hasReady || campaigned
+
+		if isLeader {
+			// The new peer is likely to become leader, send a heartbeat immediately to reduce
+			// client query miss.
+			newPeer.peer.HeartbeatPd(d.ctx)
+		}
+
+		newPeer.peer.Activate(d.ctx)
+		meta.regions[newRegionID] = newRegion
+		if lastRegionID == newRegionID {
+			// To prevent from big region, the right region needs run split
+			// check again after split.
+			newPeer.peer.SizeDiffHint = d.ctx.Cfg.RegionSplitCheckDiff
+		}
+		mb := newMailbox(sender, newPeer)
+		d.ctx.router.register(newRegionID, mb)
+		_ = d.ctx.router.send(newRegionID, NewPeerMsg(MsgTypeStart, newRegionID, nil))
+		if !campaigned {
+			if len(meta.pendingVotes) > 0 {
+				msg := meta.pendingVotes[0]
```

Comment:
I think `swap_remove_front(|m| m.get_to_peer == meta_peer)` means remove the element which satisfy `m.get_to_peer == meta_peer`.

Reply:
Seems still not correct.